src/antsibull/utils/paths.py

```diff
@@ -0,0 +1,181 @@
+# Copyright (C) 2023 Maxwell G <maxwell@gtmx.me>
+# SPDX-License-Identifier: GPL-3.0-or-later
+# GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or
+# https://www.gnu.org/licenses/gpl-3.0.txt)
+
+"""
+Utility functions for dealing with paths and directories
+"""
+
+from __future__ import annotations
+
+import contextlib
+import os
+import tempfile
+from collections.abc import AsyncIterator, Iterator
+from pathlib import Path
+from shutil import Error, copy2, copystat
+from typing import TYPE_CHECKING
+
+import aiofiles.ospath
+import aiofiles.tempfile
+
+if TYPE_CHECKING:
+    from _typeshed import StrPath
+
+
+def _is_dir(directory: Path) -> None:
+    if not directory.is_dir():
+        raise ValueError(f"{directory} is not a directory!")
+
+
+@contextlib.contextmanager
+def temp_or_dir(directory: StrPath | None = None, /) -> Iterator[Path]:
+    if directory:
+        directory = Path(directory)
+        _is_dir(directory)
+        yield directory
+    else:
+        with tempfile.TemporaryDirectory() as tmp:
+            yield Path(tmp)
+
+
+@contextlib.asynccontextmanager
+async def atemp_or_dir(directory: StrPath | None = None, /) -> AsyncIterator[Path]:
+    if directory:
+        directory = Path(directory)
+        _is_dir(directory)
+        yield directory
+    else:
+        async with aiofiles.tempfile.TemporaryDirectory() as tmp:
+            yield Path(tmp)
+
+
+def copytree_and_symlinks(
+    src: StrPath | os.DirEntry,
+    dest: StrPath | os.DirEntry,
+    *,
+    ignore_dangling_symlinks: bool = True,
+    dirs_exist_ok: bool = False,
+    _origsrc: Path | None = None,
+) -> None:
+    """
+    Modified version of `shutil.copytree()` that handles symlinks more inteligently.
+
+    * Relative symlink to files within the `src` directory -> symlink is preserved
+    * Relative symlink to files outside the `src` directory -> files are copied
+    * `ignore_dangling_symlinks` is the default.
+
+    Windows code branches are removed. Antsibull does not support Windows.
+
+    SPDX-License-Identifier: Python-2.0.1
+    Copyright (c) 2001, 2002, 2003, 2004, 2005, 2006, 2007,
+    2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019,
+    2020, 2021, 2022, 2023 Python Software Foundation; All Rights Reserved
+    """
```

Comment:
Does this info apply only to this function? I guess not, in which case it should be at the top-level.

How about moving all copied/modified code into a separate module (with a leading underscore)?